### ISSUE TITLE
Weird issue with creating custom message presenter using Chatto and ChattoAdditions with Carthage

### ISSUE THREAD

**Issue by zwang** [badoo/Chatto#60]
Hi @diegosanchezr 

I got a really weird issue using Carthage. I have spent hours on this but have no clue. I think it might be related to Carthage and Swift. I post here to see if you can help by any chance.

I have created a demo project at here: https://github.com/zwang/ChattoDiff

In the demo project, I have two simple iOS project, both created using Xcode default single page application template with one extra source file DemoMessagePresenter.swift. The two project share exactly the same source files.

However, in CApp project using Carthage for dependency management, the source code can not be compiled. In PApp project using CocoaPods for dependency management, everything compiles.

The compilation error show as below:

`/Users/zhao/workspace/ChattoDiff/Sources/DemoMessagePresenter.swift:122:38: Cannot convert value of type 'InteractionHandlerT?' to expected argument type '_?'`

The `DemoMessagePresenter` and `DemoMessagePresenterBuilder` is exactly the same as `TextMessagePresenter` and `TextMessagePresenterBuilder` in the ChattoAdditions library. I just simply changed the names

![screen shot 2016-03-13 at 11 48 12 pm](https://cloud.githubusercontent.com/assets/1171212/13737332/977f8fb4-e979-11e5-8e68-31b073a1c6c7.png)

**Comment by diegosanchezr** [comment]
Hi @zwang, this is really weird indeed. There must be something wrong with ChattoAdditions project. This workaround fixes Carthage and breaks CocoaPods. Maybe it helps to find the root issue:

```swift
diff --git a/Sources/DemoMessagePresenter.swift b/Sources/DemoMessagePresenter.swift
index a2b52fc..941bf31 100644
--- a/Sources/DemoMessagePresenter.swift
+++ b/Sources/DemoMessagePresenter.swift
@@ -7,7 +7,7 @@ public class DemoMessagePresenter<ViewModelBuilderT, InteractionHandlerT where
     ViewModelBuilderT.ViewModelT: TextMessageViewModelProtocol,
     InteractionHandlerT: BaseMessageInteractionHandlerProtocol,
     InteractionHandlerT.ViewModelT == ViewModelBuilderT.ViewModelT>
-: BaseMessagePresenter<TextBubbleView, ViewModelBuilderT, InteractionHandlerT> {
+: BaseMessagePresenter<TextBubbleView, ViewModelBuilderT, InteractionHandlerT>, ChatItemPresenterProtocol {
```

**Comment by zwang** [comment]
Hi @diegosanchezr  thank you for looking into this. I had another workaround, you can see it here: https://github.com/zwang/ChattoDiff/commit/04cc1822544a75d083667ef36d3c41eaa5972df1 

I just copied the source code of BaseMessagePresenter and put it into the same project and change the name. Then all my other custom presenter inherit from the new one. Then it works. Not ideal.

Basically make sure the base class and subclass in the same project solved the issue in a not nice way. 

I think this seems to be an issue of carthage, which leads to root cause of swift based on Carthage's known issue in Readme.

**Comment by diegosanchezr** [comment]
Hi @zwang

Looks like the issue is related to optimization level. Cocoapods by default will compile the frameworks without optimization, while Carthage will take whatever is set in the project. In other words, this error reproduces when the framework is compiled -O -whole-module-optimization and so I would say this is Apple's bug.

If someone has the time to create a bug report in bugs.swift.org please let us know. In the meanwhile feel free to make a PR changing the optimization level in Chatto/ChattoAdditions for release configuration

**Comment by zwang** [comment]
@diegosanchezr, that's good to know. It makes sense since I am not using whole-module-optimization due to some other weird issue I had before. I will verify and probably file a bug at swift.org then. Thanks for the help. :+1:

**Comment by diegosanchezr** [comment]
@zwang I had more time to look at this and create a bug report: https://bugs.swift.org/browse/SR-1158

Turns out there's a simpler workaround that works with all build configurations
```
public func createPresenterWithChatItem(chatItem: ChatItemProtocol) -> ChatItemPresenterProtocol {
    assert(self.canHandleChatItem(chatItem))
    let presenter =  DemoMessagePresenter<ViewModelBuilderT, InteractionHandlerT>(
        messageModel: chatItem as! ModelT,
        viewModelBuilder: self.viewModelBuilder,
        interactionHandler: self.interactionHandler,
        sizingCell: sizingCell,
        baseCellStyle: self.baseMessageStyle,
        textCellStyle: self.textCellStyle,
        layoutCache: self.layoutCache
    )
    return presenter as! ChatItemPresenterProtocol // https://bugs.swift.org/browse/SR-1158
}
```

**Comment by zwang** [comment]
@diegosanchezr Thank you for looking into this more and filing the bug. The work around seems to be much simpler. :+1:

**Comment by diegosanchezr** [comment]
Fixed in Xcode 7.3.1
